src/environment.rs

```diff
@@ -354,32 +391,68 @@ impl Environment {
         self.cpu.regs_mut()[cpu::Cpu::SP] = stack_high_addr;
         self.cpu.regs_mut()[0] = user_data.to_bits();
         self.cpu
-            .branch_with_link(start_routine, self.dyld.return_to_host_routine());
+            .branch_with_link(start_routine, self.dyld.thread_exit_routine());
         self.switch_thread(old_thread);
 
         new_thread_id
     }
 
     /// Put the current thread to sleep for some duration, running other threads
-    /// in the meantime as appropriate.
-    pub fn sleep(&mut self, duration: Duration) {
-        assert!(self.threads[self.current_thread].sleeping_until.is_none());
+    /// in the meantime as appropriate. Functions that call sleep right before they return should set
```

Comment:
This should be qualified: before returning to the guest/`Environment::run`. Functions that can be called directly by other host functions shouldn't use this.